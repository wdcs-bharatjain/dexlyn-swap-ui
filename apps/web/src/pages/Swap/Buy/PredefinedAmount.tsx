import styled, { css } from 'styled-components'
import { useSporeColors } from 'ui/src'
import { Pill } from 'uniswap/src/components/pill/Pill'
import { NO_DECIMALS_CURRENCY, useFormatter } from 'utils/formatNumbers'

interface PredefinedAmountProps {
  amount: number
  currentAmount: string
  disabled?: boolean
  onClick: () => void
}

const ClickablePill = styled(Pill)<{ disabled: boolean }>`
  ${({ disabled }) =>
    !disabled &&
    css`
      cursor: pointer;
      &:hover {
        background-color: ${({ theme }) => theme.surface2};
        border-color: ${({ theme }) => theme.surface3};
      }
    `}
`

export function PredefinedAmount({ currentAmount, amount, disabled = false, onClick }: PredefinedAmountProps) {
  const colors = useSporeColors()
  const { formatFiatPrice } = useFormatter()

  const highlighted = currentAmount === amount.toString()
  return (
    <ClickablePill
      disabled={disabled}
      onPress={onClick}
      backgroundColor={!disabled && highlighted ? '$surface2' : '$surface1'}
      customBorderColor={disabled ? colors.surface2.val : colors.surface3.val}
      foregroundColor={colors[disabled ? 'neutral3' : highlighted ? 'neutral1' : 'neutral2'].val}
      label={formatFiatPrice({
        price: amount,
        type: [
          {
            upperBound: Number.MAX_SAFE_INTEGER,
            formatterOptions: { ...NO_DECIMALS_CURRENCY, useGrouping: false },
          },
        ],
      })}
      px="$spacing16"
      textVariant="buttonLabel3"
    />
  )
}
