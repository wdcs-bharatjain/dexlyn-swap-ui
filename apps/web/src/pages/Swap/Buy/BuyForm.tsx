import { ChainId } from '@uniswap/sdk-core'
import { useAccountDrawer } from 'components/AccountDrawer/MiniPortfolio/hooks'
import { ButtonLight, ButtonPrimary } from 'components/Button'
import Column from 'components/Column'
import Row from 'components/Row'
import { buildCurrencyInfo } from 'constants/routing'
import { nativeOnChain } from 'constants/tokens'
import { useAccount } from 'hooks/useAccount'
import { useActiveLocalCurrencyComponents } from 'hooks/useActiveLocalCurrency'
import { useUSDTokenUpdater } from 'hooks/useUSDTokenUpdater'
import { PredefinedAmount } from 'pages/Swap/Buy/PredefinedAmount'
import {
  NumericalInputMimic,
  NumericalInputSymbolContainer,
  NumericalInputWrapper,
  StyledNumericalInput,
  useWidthAdjustedDisplayValue,
} from 'pages/Swap/common/shared'
import { useCallback, useMemo, useState } from 'react'
import { Trans } from 'react-i18next'
import styled from 'styled-components'
import { Text } from 'ui/src/components/text/Text'
import { FiatOnRampCountryPicker } from 'uniswap/src/features/fiatOnRamp/FiatOnRampCountryPicker'
import { SelectTokenButton } from 'uniswap/src/features/fiatOnRamp/SelectTokenButton'
import { FiatOnRampCurrency } from 'uniswap/src/features/fiatOnRamp/types'
import Trace from 'uniswap/src/features/telemetry/Trace'
import { InterfacePageNameLocal } from 'uniswap/src/features/telemetry/constants'
import useResizeObserver from 'use-resize-observer'
import { NumberType, useFormatter } from 'utils/formatNumbers'

const InputWrapper = styled(Column)`
  position: relative;
  background-color: ${({ theme }) => theme.surface2};
  padding: 16px;
  height: 312px;
  align-items: center;
  border-radius: 20px;
  justify-content: center;
  overflow: hidden;
  gap: 16px;
`

const HeaderRow = styled(Row)`
  align-items: center;
  justify-content: space-between;
  position: absolute;
  top: 16px;
  left: 16px;
  width: calc(100% - 32px);
`

const PREDEFINED_AMOUNTS = [100, 300, 1000]
const ethCurrencyInfo = buildCurrencyInfo(nativeOnChain(ChainId.MAINNET))

type BuyFormProps = {
  disabled?: boolean
}

export function BuyForm({ disabled }: BuyFormProps) {
  const { isConnected } = useAccount()
  const accountDrawer = useAccountDrawer()
  const { formatNumberOrString } = useFormatter()
  const { symbol: fiatSymbol } = useActiveLocalCurrencyComponents()
  const [inputAmount, setInputAmount] = useState<string>('')
  const [quoteCurrency, setQuoteCurrency] = useState<FiatOnRampCurrency>({
    currencyInfo: ethCurrencyInfo,
    meldCurrencyCode: 'ETH',
  })
  const postWidthAdjustedDisplayValue = useWidthAdjustedDisplayValue(inputAmount)
  const hiddenObserver = useResizeObserver<HTMLElement>()
  const exactAmountOut = useUSDTokenUpdater(true /* inputInFiat */, inputAmount, quoteCurrency?.currencyInfo?.currency)
  const buyButtonState = useMemo(() => {
    if (!isConnected) {
      return {
        label: <Trans i18nKey="common.connectWallet.button" />,
        disabled: false,
        onClick: accountDrawer.open,
        Component: ButtonLight,
      }
    }
    if (!inputAmount) {
      return {
        label: <Trans i18nKey="common.noAmount.error" />,
        disabled: true,
        onClick: undefined,
        Component: ButtonPrimary,
      }
    }

    return {
      label: <Trans i18nKey="common.continue.button" />,
      disabled: false,
      Component: ButtonPrimary,
      onClick: () => {
        // TODO: open "choose FOR provider" modal
      },
    }
  }, [accountDrawer.open, inputAmount, isConnected])

  const handleUserInput = useCallback((newValue: string) => {
    setInputAmount(newValue)
  }, [])

  return (
    <Trace page={InterfacePageNameLocal.Buy}>
      <Column gap="xs">
        <InputWrapper>
          <HeaderRow>
            <Text variant="body3" userSelect="none" color="$neutral2">
              <Trans i18nKey="common.youreBuying" />
            </Text>
            <FiatOnRampCountryPicker
              onPress={function (): void {
                // TODO: open country picker modal
              }}
              // TODO: get country code from user's location
              countryCode="US"
            />
          </HeaderRow>
          <NumericalInputWrapper>
            <NumericalInputSymbolContainer showPlaceholder={!inputAmount}>{fiatSymbol}</NumericalInputSymbolContainer>
            <StyledNumericalInput
              value={postWidthAdjustedDisplayValue}
              disabled={disabled}
              onUserInput={handleUserInput}
              placeholder="0"
              $width={inputAmount && hiddenObserver.width ? hiddenObserver.width + 1 : undefined}
              maxDecimals={6}
            />
            <NumericalInputMimic ref={hiddenObserver.ref}>{inputAmount}</NumericalInputMimic>
          </NumericalInputWrapper>
          <SelectTokenButton
            onPress={() => {
              // TODO: open currency selector modal and call setQuoteCurrency
              setQuoteCurrency({
                currencyInfo: ethCurrencyInfo,
                meldCurrencyCode: 'ETH',
              })
            }}
            selectedCurrencyInfo={quoteCurrency.currencyInfo ?? ethCurrencyInfo}
            formattedAmount={formatNumberOrString({
              input: exactAmountOut || '0',
              type: NumberType.TokenNonTx,
            })}
            showCaret={false}
            disabled={disabled}
          />
          <Row gap="md" justify="center">
            {PREDEFINED_AMOUNTS.map((amount: number) => (
              <PredefinedAmount
                onClick={() => {
                  setInputAmount(amount.toString())
                }}
                key={amount}
                amount={amount}
                currentAmount={inputAmount}
                disabled={disabled}
              />
            ))}
          </Row>
        </InputWrapper>
        <buyButtonState.Component fontWeight={535} disabled={buyButtonState.disabled} onClick={buyButtonState.onClick}>
          {buyButtonState.label}
        </buyButtonState.Component>
      </Column>
    </Trace>
  )
}
