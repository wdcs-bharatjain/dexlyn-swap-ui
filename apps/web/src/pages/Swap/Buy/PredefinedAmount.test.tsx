import { PredefinedAmount } from 'pages/Swap/Buy/PredefinedAmount'
import { render, screen } from 'test-utils/render'

describe('PredefinedAmount', () => {
  it.each([
    [100, '1000', false],
    [300, '1000', true],
    [1000, '1000', false],
    [1000, '1000', true],
  ])(
    'renders correctly with amount= %p , currentAmount= %p , disabled= %p',
    async (amount, currentAmount, disabled) => {
      const clickHandler = jest.fn()
      const { container } = render(
        <PredefinedAmount amount={amount} currentAmount={currentAmount} disabled={disabled} onClick={clickHandler} />
      )
      screen.getByText('$' + amount).click()
      expect(clickHandler).toHaveBeenCalledTimes(disabled ? 0 : 1)

      expect(container.firstChild).toMatchSnapshot()
    }
  )
})
