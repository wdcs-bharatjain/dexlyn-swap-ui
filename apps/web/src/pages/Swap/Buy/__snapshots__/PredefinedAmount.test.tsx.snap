// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PredefinedAmount renders correctly with amount= 100 , currentAmount= "1000" , disabled= false 1`] = `
.c0 {
  cursor: pointer;
}

.c0:hover {
  background-color: #F9F9F9;
  border-color: #22222212;
}

<span
  class="t_light is_inversed _dsp_contents"
>
  <span
    class=" t_light _dsp_contents is_Theme"
  >
    <div
      class="_display-flex _flexBasis-auto _boxSizing-border-box _position-relative _minHeight-0px _minWidth-0px _flexShrink-0 _alignItems-center _borderTopLeftRadius-1307610122 _borderTopRightRadius-1307610122 _borderBottomRightRadius-1307610122 _borderBottomLeftRadius-1307610122 _borderTopWidth-1px _borderRightWidth-1px _borderBottomWidth-1px _borderLeftWidth-1px _flexDirection-row _gap-1481558276 _justifyContent-center _pr-1481558214 _pl-1481558214 _pt-1481558276 _pb-1481558276 _backgroundColor-568007324 _borderBottomStyle-solid _borderTopStyle-solid _borderLeftStyle-solid _borderRightStyle-solid c0"
      style="border-color: #2222220d;"
    >
      <span
        class="font_button _fontFamily-299667014 _display-inline _boxSizing-border-box _wordWrap-break-word _whiteSpace-pre-wrap _mt-0px _mr-0px _mb-0px _ml-0px _color-843134974 _fontSize-16px _lineHeight-24px _fontWeight-500"
        data-disable-theme="true"
        style="color: rgb(125, 125, 125); padding-top: 1px;"
      >
        $100
      </span>
    </div>
  </span>
</span>
`;

exports[`PredefinedAmount renders correctly with amount= 300 , currentAmount= "1000" , disabled= true 1`] = `
<span
  class=" _dsp_contents"
>
  <span
    class=" t_light _dsp_contents is_Theme"
  >
    <div
      aria-disabled="true"
      class="_display-flex _flexBasis-auto _boxSizing-border-box _position-relative _minHeight-0px _minWidth-0px _flexShrink-0 _alignItems-center _borderTopLeftRadius-1307610122 _borderTopRightRadius-1307610122 _borderBottomRightRadius-1307610122 _borderBottomLeftRadius-1307610122 _borderTopWidth-1px _borderRightWidth-1px _borderBottomWidth-1px _borderLeftWidth-1px _flexDirection-row _gap-1481558276 _justifyContent-center _pr-1481558214 _pl-1481558214 _pt-1481558276 _pb-1481558276 _backgroundColor-568007324 _borderBottomStyle-solid _borderTopStyle-solid _borderLeftStyle-solid _borderRightStyle-solid "
      style="border-color: #f9f9f9;"
    >
      <span
        class="font_button _fontFamily-299667014 _display-inline _boxSizing-border-box _wordWrap-break-word _whiteSpace-pre-wrap _mt-0px _mr-0px _mb-0px _ml-0px _color-843134974 _fontSize-16px _lineHeight-24px _fontWeight-500"
        data-disable-theme="true"
        style="color: rgb(206, 206, 206); padding-top: 1px;"
      >
        $300
      </span>
    </div>
  </span>
</span>
`;

exports[`PredefinedAmount renders correctly with amount= 1000 , currentAmount= "1000" , disabled= false 1`] = `
.c0 {
  cursor: pointer;
}

.c0:hover {
  background-color: #F9F9F9;
  border-color: #22222212;
}

<span
  class=" _dsp_contents"
>
  <span
    class=" t_light _dsp_contents is_Theme"
  >
    <div
      class="_display-flex _flexBasis-auto _boxSizing-border-box _position-relative _minHeight-0px _minWidth-0px _flexShrink-0 _alignItems-center _borderTopLeftRadius-1307610122 _borderTopRightRadius-1307610122 _borderBottomRightRadius-1307610122 _borderBottomLeftRadius-1307610122 _borderTopWidth-1px _borderRightWidth-1px _borderBottomWidth-1px _borderLeftWidth-1px _flexDirection-row _gap-1481558276 _justifyContent-center _pr-1481558214 _pl-1481558214 _pt-1481558276 _pb-1481558276 _backgroundColor-568007293 _borderBottomStyle-solid _borderTopStyle-solid _borderLeftStyle-solid _borderRightStyle-solid c0"
      style="border-color: #2222220d;"
    >
      <span
        class="font_button _fontFamily-299667014 _display-inline _boxSizing-border-box _wordWrap-break-word _whiteSpace-pre-wrap _mt-0px _mr-0px _mb-0px _ml-0px _color-843134974 _fontSize-16px _lineHeight-24px _fontWeight-500"
        data-disable-theme="true"
        style="color: rgb(34, 34, 34); padding-top: 1px;"
      >
        $1000
      </span>
    </div>
  </span>
</span>
`;

exports[`PredefinedAmount renders correctly with amount= 1000 , currentAmount= "1000" , disabled= true 1`] = `
<span
  class=" _dsp_contents"
>
  <span
    class=" t_light _dsp_contents is_Theme"
  >
    <div
      aria-disabled="true"
      class="_display-flex _flexBasis-auto _boxSizing-border-box _position-relative _minHeight-0px _minWidth-0px _flexShrink-0 _alignItems-center _borderTopLeftRadius-1307610122 _borderTopRightRadius-1307610122 _borderBottomRightRadius-1307610122 _borderBottomLeftRadius-1307610122 _borderTopWidth-1px _borderRightWidth-1px _borderBottomWidth-1px _borderLeftWidth-1px _flexDirection-row _gap-1481558276 _justifyContent-center _pr-1481558214 _pl-1481558214 _pt-1481558276 _pb-1481558276 _backgroundColor-568007324 _borderBottomStyle-solid _borderTopStyle-solid _borderLeftStyle-solid _borderRightStyle-solid "
      style="border-color: #f9f9f9;"
    >
      <span
        class="font_button _fontFamily-299667014 _display-inline _boxSizing-border-box _wordWrap-break-word _whiteSpace-pre-wrap _mt-0px _mr-0px _mb-0px _ml-0px _color-843134974 _fontSize-16px _lineHeight-24px _fontWeight-500"
        data-disable-theme="true"
        style="color: rgb(206, 206, 206); padding-top: 1px;"
      >
        $1000
      </span>
    </div>
  </span>
</span>
`;
